cpp/src/arrow/testing/gmock.h

```diff
@@ -0,0 +1,31 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <gmock/gmock.h>
+
+namespace arrow {
+
+using testing::Eq;
+using testing::HasSubstr;
+
+MATCHER_P(Equals, other, "") { return arg.Equals(other); }
+MATCHER_P(PtrEquals, other, "") { return arg->Equals(*other); }
+MATCHER_P(OkAndEq, other, "") { return arg.ok() && arg.ValueOrDie() == other; }
```

Comment:
I admit I fail to understand the point of these compared to the existing available assertions (such as `ASSERT_OK_AND_EQ` or the various `Assert*Equals` helpers).

Reply:
I think this is needed to be able to compose with the gmock APIs